### DISCUSSION TITLE
Deploy a project within a monorepo with non-symlinked node_modules

### DISCUSSION THREAD

**Discussion by koalazub** [pnpm/pnpm#5609] [Q&A]
Hey! Before I go on a massive spewl regarding the details, I thought to pin the aim at the very top:
- Deployment of a project within a pnpm workspace with its dependencies into a lambda

I'm wanting to deploy a module within a monorepo I have in pnpm. I don't know enough just yet as I'm in the middle of migrating our work and so if I get the terminology wrong then please correct(I'll try hard to reedit where I think I'm vague).
I was looking at `pnpm deploy`: https://github.com/pnpm/pnpm/issues/4378 to maybe help, but I'm unsure exactly as to what it does. The unknowns for me is that it still uses symlinks and that won't work if I deploy in isolation from the workspace

The second is using dedicated lockfiles: https://github.com/pnpm/pnpm/issues/1919
Is this a viable method to do what I'm aiming to do. And because I'm still unsure of everything, is there a better example to reference when implementing this step? Does it actually help me create tangible node_modules that I'm able to export along with my built project to a lambda

The final Issue I looked into: https://github.com/pnpm/pnpm/issues/2198 points out that I can use PnP:https://github.com/pnpm/pnpm/issues/2198#issuecomment-710882357 
My problem with this step is whether this is only for the monorepo as a whole, or if I could use it in isolation for that singular project to deploy into a lambda.

Tl;Dr 
- Tried --shamefully-hoist to replace the symlinks with node_modules for a project
- Tried using pnpm deploy and still got symlinked node_modules
- Tried looking into whether others have figured out how to extract one project from their monorepo and haven't been able to do this on my end

My monorepo consists of 4 node projects, with one of the repos depending on several of the other packages in that repo. I use it and deploy it to a lambda for use. I'm looking for help with conceptualising what pnpm is exactly doing during the build/deploy steps because of the means of symlinking and thus me reaching out for a bit help in this regard. I hope that helps explain what I'm attempting, and let me know if there's a gap in what I've said

**Comment by lvqq** [comment]
Lambda doesn't support symlink, you can try `node-linker= hoisted`: https://pnpm.io/npmrc#node-linker

**Reply by koalazub** [reply]
As in we'd need to add this to the `npmrc` file directly, or to just set the config via the `pnpm cli` command before runniing  `deploy`? Or either works?

So:

```
; .npmrc
  
registry=blah
@scopecompany:registry=blah.com/
//npm.pkg.github.com=token

node-linker = hoisted pnpm
```

Is that right?

**Reply by lvqq** [reply]
Either works, but I recommand the first way. `pnpm/npm config` update the user config by default, it's better to config under the project.

**Reply by vizio360** [reply]
I did ask a similar question 2 hours before you posted this :). https://github.com/pnpm/pnpm/discussions/5608

Same for me, I would like to release a service from a monorepo with all dependencies in it's own node_module. Tbh I don't mind if they are still sym linked within the node_modules folder, but there should be no references to the root of the repo or symlinks to local packages, all dependencies should be downloaded from the registry so to have a portable standalone installation of the service.
I'll try this `node-linker` option as well.

**Comment by koalazub** [comment]
Posting this [pnpm Discord discussion here for visbility](https://discord.com/channels/731599538665553971/747905674578624633/1039852953110052975)

**Reply by koalazub** [reply]
The summary is that you can use a bundler to assist with the issue of node modules such as [aws cdk](https://docs.aws.amazon.com/cdk/v2/guide/serverless_example.html), [turbopack](https://turbo.build/pack) etc.
